learning_phase/chat_sample/client.go

```diff
@@ -0,0 +1,88 @@
+package main
+
+import (
+	"bufio"
+	"context"
+	"fmt"
+	"log"
+	"net"
+	"os"
+	"strings"
+)
+
+//various commands numbered from 0, 1, 2, 3...
+const (
+	cmdauth = iota
+	cmdpm
+	cmdmsg
+	cmdquit
+)
+
+type command struct {
+	id     int
+	client *client
+	args   []string
+}
+
+type client struct {
+	conn     net.Conn
+	address  string
+	username string
+	commands chan<- command //incoming channel
+}
+
+func (c *client) readInput(ctx context.Context) {
+	for {
+		select {
+		case <-ctx.Done(): //when cancel is called
+			log.Printf("Terminating Clients")
+			os.Exit(0)
+		default:
+			msg, err := bufio.NewReader(c.conn).ReadString('\n')
```

Comment:
TCP guarantees only reliability and ordering of message delivery, and not that the entire message will be delivered in one segment. This statement may read only a part of the message received, without actually waiting to receive the entire message. Read this https://stackoverflow.com/a/47585913 for a better idea